jax/interpreters/pxla.py

```diff
@@ -237,17 +237,17 @@ def shard_args(devices: Sequence[xb.xla_client.Device],
 
 shard_arg_handlers: Dict[Any, Callable[[Any, Any, Any], Sequence[Any]]] = {}
 shard_arg_handlers[core.Unit] = \
-    lambda x, devices, _: [xla.device_put(core.unit, d) for d in devices]
+    lambda x, devices, _: list(it.chain.from_iterable(xla.device_put(core.unit, d) for d in devices))
 def _shard_array(x, devices, indices):
-  return [xla.device_put(x[i], d) for (i, d) in zip(indices, devices)]
+  return list(it.chain.from_iterable(xla.device_put(x[i], d) for (i, d) in zip(indices, devices)))
```

Comment:
FWIW I did some basic timings, first to compare writing this flattening with `it.chain.from_iterable` versus with a list comprehension:

```
In [5]: xs = [(1, 1) for _ in range(2)]

In [6]: ys = [(1, 1) for _ in range(2000)]

In [7]: timeit list(it.chain.from_iterable(x for x in xs))
462 ns ± 3.69 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [8]: timeit [x for pair in xs for x in pair]
214 ns ± 2.67 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [9]: timeit list(it.chain.from_iterable(x for x in ys))
94 µs ± 92.8 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [10]: timeit [y for pair in ys for y in pair]
104 µs ± 2.16 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

That suggests to me that writing this as `it.chain.from_iterable` is probably preferable to writing it as a list comprehension, though the latter might be slightly more readable.

However, comparing this to a version that doesn't have to do any flattening:

```
In [11]: xs = [1 for _ in range(2)]

In [12]: ys = [1 for _ in range(2000)]

In [13]: timeit [x for x in xs]
134 ns ± 1.12 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [14]: timeit [y for y in ys]
28.4 µs ± 37.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Comparing this 28.4µs to the 94µs number makes me a _little_ concerned that we're adding overhead to this dispatch path. Maybe this dispatch path isn't the important one for pxla.py though; the one we care about is handling ShardedDeviceArrays, which we do directly in `shard_args` and which wasn't touched by this PR.

So I _think_ this is all good, but I wanted to raise the issue for discussion just in case. WDYT @jakevdp @skye @hawkinsp ? Any concern about overheads here?

Reply:
I ran some of the other scripts in the `benchmarks` folder, and it seems to be a mixed bag: some things slower, some things faster. It's hard for me to tell from the printouts whether any of the reported differences are significant.

Reply:
Could you add that info to the PR message? That's good for posterity, but also good for others to weigh in on.

Reply:
Done, see below.